### ISSUE TITLE
S1030: don't flag m[string(buf.Bytes()]

### ISSUE THREAD

**Issue by bep** [dominikh/go-tools#723]
https://staticcheck.io/docs/checks#S1030

>bytes.Buffer has both a String and a Bytes method. It is never necessary to use string(buf.Bytes()) or []byte(buf.String()) – simply use the other method.

While this may have been correct in older versions of Go, have a look at this benchmark:

```go
func BenchmarkBytesVsString(b *testing.B) {
	seen := make(map[string]bool)
	var buff bytes.Buffer
	buff.WriteString(strings.Repeat("hello ", 100))

	b.Run("buff.String()", func(b *testing.B) {
		for i := 0; i < b.N; i++ {
			if seen[buff.String()] {
			}
		}
	})

	b.Run("buff.Bytes()", func(b *testing.B) {
		for i := 0; i < b.N; i++ {
			if seen[string(buff.Bytes())] {
			}
		}
	})

}

```

Running it:

```bash
BenchmarkBytesVsString/buff.String()-16         	 9718815	       110 ns/op	     640 B/op	       1 allocs/op
BenchmarkBytesVsString/buff.Bytes()-16          	456494642	         2.56 ns/op	       0 B/op	       0 allocs/op
```
From what I've read, this is a special optimisation for map lookups.

**Comment by ainar-g** [comment]
<p>
Interesting.  Seems like the branch in <code>String</code> makes it less
optimisable in <i>go tool compile</i>.  I've made a slightly different
version of the benchmark using
<a
href="https://dave.cheney.net/2013/06/30/how-to-write-benchmarks-in-go">Dave Cheney's recomendations</a>:
</p>

``` go
package main

import (
	"bytes"
	"strings"
	"testing"
)

var sink bool

func BenchmarkBytesVsString(b *testing.B) {
	var buff bytes.Buffer
	buff.WriteString(strings.Repeat("hello ", 100))

	var seen = make(map[string]bool)
	b.Run("buff.String()", func(b *testing.B) {
		var lsink bool
		for i := 0; i < b.N; i++ {
			lsink = seen[buff.String()]
		}

		sink = lsink
	})

	b.Run("string(buff.Bytes())", func(b *testing.B) {
		var lsink bool
		for i := 0; i < b.N; i++ {
			lsink = seen[string(buff.Bytes())]
		}

		sink = lsink
	})
}
```

<p>
Also, the results are different when using <i>gccgo</i>:
</p>

```
$ go test --bench=. --compiler=gccgo --gccgoflags="-O2" ./tmp_test.go
goos: linux
goarch: amd64
BenchmarkBytesVsString/buff.String()-4         	 3000000	       474 ns/op
BenchmarkBytesVsString/string(buff.Bytes())-4  	 3000000	       473 ns/op
PASS
ok  	command-line-arguments	3.938s
```

**Comment by bep** [comment]
@ainar-g I'm confused, are you saying my numbers are wrong?

**Comment by ainar-g** [comment]
<p>
No, no, I'm just adding data points.  Your numbers are correct for <i>go
tool compile</i>, the default compiler.  It's up to Dominik to decide
what to do with the check.
</p>

**Comment by bep** [comment]
OK. If you search the git commit log/issue tracker you'll find the compiler optimisation for this. I think it's in Go >= 1.13  -- and it's for map lookups only, and when using the `mymap[string(mybytes)]` construct -- which will not allocate for the string construct.

**Comment by dominikh** [comment]
Go (for quite some time now, long before 1.13) optimizes `m[string(b)]` where `b []byte` – it avoids copying the underlying data, which is normally required for conversions from `[]byte` to `string`. I don't know if this was ever the case for `m[buf.String()]` where `b *bytes.Buffer`. I would not think so.

S1030 shouldn't make the suggestion in the presence of this pattern.

**Comment by dominikh** [comment]
Closed by 87cefbaa37b829d7c336e580e22527d97604c941.
